src/sqlfluff/core/linter/runner.py

```diff
@@ -127,54 +133,128 @@ def __init__(self, linter, config, processes) -> None:
         super().__init__(linter, config)
         self.processes = processes
 
-    def run(self, fnames: List[str], fix: bool):
+    def run(self, fnames: List[str], fix: bool) -> Iterator[LintedFile]:
         """Parallel implementation.
 
-        Note that the partials are generated one at a time then
-        passed directly into the pool as they're ready. This means
-        the main thread can do the IO work while passing the parsing
+        Partials are generated one by one. If a file has been successfully
+        rendered, it is set up as an asynchronous job, and if it's been
+        unsuccessful we handle it immediately in the main process. This
+        means the main thread can do the IO work while passing the parsing
         and linting work out to the threads.
+
+        Any completed jobs are always handled before queueing more jobs,
+        and the queue is limited in size, which when reached we simply
+        wait for some progress to be made before enqueueing more work.
+
+        We keep track of the outstanding jobs by keeping a reference to
+        the AsyncResult of each, which is deleted upon receipt of a
+        result from that job.
         """
         with self._create_pool(
             self.processes,
             self._init_global,
             (self.config,),
         ) as pool:
             try:
-                for lint_result in self._map(
-                    pool,
-                    self._apply,
-                    self.iter_partials(fnames, fix=fix),
-                ):
-                    if isinstance(lint_result, DelayedException):
+                # Make a queue to hold the results.
+                queue: multiprocessing.Queue[Union[DelayedException, LintedFile]] = (
+                    multiprocessing.Queue()
+                )
+                # And a dict to hold the async references
+                async_map: Dict[str, multiprocessing.pool.AsyncResult] = {}
+                # Target queue size should be twice the the number of processes.
+                target_queue_depth = self.processes * 2
+                partials_iterable = self.iter_partials(fnames, fix=fix)
+
+                while True:
+                    # If there's something to process, do so.
+                    try:
+                        lint_result = queue.get(block=False)
```

Comment:
Should we at least sleep for a brief period? Otherwise, it looks like this loop would peg one of the CPUs at 100%.